src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,382 @@
+            new_stroke.push_back(new_point);
+        }
+        // Apply RDP Line Algorithm
+        flex_list compressed_stroke = ramer_douglas_peucker(new_stroke, 2.0);
+        simplified_drawing.push_back(compressed_stroke);
+    }
+    return simplified_drawing;
+}
+
+void swap_points(Point *P1, Point *P2) {
+    float temp_x, temp_y;
+    if (P1->x > P2->x) {
+        temp_x = P1->x;
+        temp_y = P1->y;
+        P1->x = P2->x;
+        P1->y = P2->y;
+        P2->x = temp_x;
+        P2->y = temp_y;
+    }
+}
+
+bool in_bounds(int x, int y, int dim) {
+    return (x >= 0 && x < dim && y >= 0 && y < dim);
+}
+
+flex_nd_vec paint_stroke(
+    flex_nd_vec bitmap, Point start, Point end, float stroke_width) {
+    swap_points(&start, &end);
+    float slope = (end.y - start.y)/(end.x - start.x);
+    int x1 = (int)(start.x);
+    int y1 = (int)(start.y);
+    int x2 = (int)(end.x);
+    int y2 = (int)(end.y);
+    int pad = (int)(stroke_width/2);
+    size_t dimension = bitmap.shape()[1];
+    if (x1 == x2) {
+        // just paint the y axis
+        int min_y = MIN(y1, y2);
+        int max_y = MAX(y1, y2);
+        for (int y = min_y; y < max_y; y++) {
+            for (int dx = -pad; dx < pad; dx++) {
+                for (int dy = -pad; dy < pad; dy++) {
+                    // figure out how to do this with the flex_list
+                    if (in_bounds(x1+dx, y+dy, dimension)) {
+                        bitmap[(y+dy)*dimension + (x1+dx)] = 1.0;
+                    }
+                }
+            }
+        }
+    } else {
+        for (int x = x1; x < x2; x++) {
+            int y = (int)(slope * (x - x1) + y1);
+            for (int dx = -pad; dx < pad; dx++) {
+                for (int dy = -pad; dy < pad; dy++) {
+                    if (in_bounds(x+dx, y+dy, dimension)) {
+                        bitmap[(y+dy) * dimension + (x+dx)] = 1.0;
+                    }
+                }
+            }
+        }
+    }
+    return bitmap;
+}
+
+flex_image blur_bitmap(flex_nd_vec bitmap, int ksize) {
+    std::vector<size_t> bitmap_shape = bitmap.shape();
+    flex_nd_vec blurred_bitmap(bitmap_shape, 0.0);
+    int dimension = bitmap_shape[1];
+    int pad = ksize/2;
+    for (int row = 0; row < dimension; row++) {
+        for (int col = 0; col < dimension; col++) {
+            int index = row * dimension + col;
+            if (row < pad 
+                || row >= dimension-pad 
+                || col < pad 
+                || col >= dimension-pad) {
+                blurred_bitmap[index] = (int)(255*bitmap[index]);
+                continue;
+            }
+            double sum_for_blur = 0.0;
+            for (int dr = -pad; dr <= pad; dr++) {
+                for (int dc = -pad; dc <= pad; dc++) {
+                    sum_for_blur += bitmap[(row+dr) * dimension + (col+dc)];
+                }
+            }
+            blurred_bitmap[index] = (int)(255*sum_for_blur/(ksize*ksize));
```

Comment:
You should guard again assigning values greater than 255 here.

Note that if `ksize` is 2, then `pad` is 1, and `ksize*ksize` is 4 but you're adding 9 values here.

Reply:
Huh, why are you casting to int here? `flex_nd_vec` stores doubles